File: python/tn/identity.py
"""TN per-user identity: XDG-based identity.json, BIP-39 bedrock, HKDF chain.

This is the module every verb in the wallet story funnels through:

    tn init <project>        -> Identity.create_new() + ensure_written()
    tn wallet restore        -> Identity.from_mnemonic(words) + ensure_written()
    tn.init(yaml_path)       -> Identity.load_or_ephemeral(...)
    tn wallet link           -> Identity.load() -> link a ceremony -> write

The HKDF ``info`` strings below are part of the stable ABI. Changing them
breaks every existing user's recovery. Don't touch them unless you
understand that.

Public helpers
--------------
``_resolve_did_endpoint(did_str)``
    Derive the HTTP base URL for a vault handler from a DID string.
    Supports ``did:key`` (falls back to ``TN_VAULT_DEFAULT_BASE`` env var)
    and ``did:web`` (fetches the DID document once and pins the endpoint).
"""

from __future__ import annotations

import base64
import json
import logging
import os
import sys
from dataclasses import asdict, dataclass, field
from pathlib import Path

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from mnemonic import Mnemonic

_log = logging.getLogger("tn.identity")

IDENTITY_SCHEMA_VERSION = 1

# --- HKDF info strings: stable ABI, never change -------------------------

HKDF_SALT = b"tn:v1"
HKDF_INFO_ROOT = b"tn:root:v1"
HKDF_INFO_DEVICE = b"tn:device:v1"
HKDF_INFO_VAULT_WRAP = b"tn:vault:wrap:v1"
HKDF_INFO_PASSKEY_SEED_WRAP = b"tn:passkey:seed-wrap:v1"


def _hkdf(ikm: bytes, info: bytes, length: int = 32, salt: bytes = HKDF_SALT) -> bytes:
    return HKDF(
        algorithm=hashes.SHA256(),
        length=length,
        salt=salt,
        info=info,
    ).derive(ikm)


def _did_key_from_ed25519_pub(pub_bytes: bytes) -> str:
    """Format the Ed25519 public key as a did:key string.

    Uses the standard multicodec prefix 0xED 0x01 for ed25519-pub then
    multibase base58btc 'z' encoding. Mirrors what tn/signing.py does
    so the resulting DIDs interop with the rest of the SDK.
    """
    # 0xed01 = ed25519-pub multicodec
    prefixed = b"\xed\x01" + pub_bytes
    alphabet = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    # base58 encode
    n = int.from_bytes(prefixed, "big")
    out = b""
    while n > 0:
        n, r = divmod(n, 58)
        out = alphabet[r : r + 1] + out
    # preserve leading zero bytes as '1' chars
    pad = 0
    for b in prefixed:
        if b == 0:
            pad += 1
        else:
            break
    return "did:key:z" + ("1" * pad + out.decode("ascii"))


# --- Paths ----------------------------------------------------------------


def _default_identity_dir() -> Path:
    """Return the XDG (or Windows equivalent) identity directory.

    Order:
      1. TN_IDENTITY_DIR if set (explicit override, cross-platform)
      2. XDG_DATA_HOME if set (POSIX + explicit on Windows)
      3. Platform default: ~/.local/share/tn on POSIX,
         %APPDATA%\\tn on Windows
    """
    override = os.environ.get("TN_IDENTITY_DIR")
    if override:
        return Path(override)
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / "tn"
    if sys.platform == "win32":
        base = os.environ.get("APPDATA") or str(Path.home() / "AppData" / "Roaming")
        return Path(base) / "tn"
    return Path.home() / ".local" / "share" / "tn"


def _default_identity_path() -> Path:
    return _default_identity_dir() / "identity.json"


# --- Identity -------------------------------------------------------------


@dataclass
class IdentityPrefs:
    default_new_ceremony_mode: str = "local"  # "local" | "linked"

    @classmethod
    def from_dict(cls, d: dict | None) -> IdentityPrefs:
        d = d or {}
        return cls(
            default_new_ceremony_mode=d.get(
                "default_new_ceremony_mode",
                "local",
            ),
        )


class IdentityError(RuntimeError):
    """Raised when identity.json is missing, corrupt, or incompatible."""


@dataclass
class Identity:
    """The user's per-machine identity.

    Loaded from or written to `$XDG_DATA_HOME/tn/identity.json` (or the
    Windows equivalent). Contains the DID, the device private key in
    the form tn/signing.py expects, the BIP-39 seed (so vault_wrap_key
    is derivable without re-typing the mnemonic every CLI call),
    optional mnemonic words (held in memory only for display), the
    linked vault URL if any, and a local cache of account-wide prefs.
    """

    did: str
    device_pub_b64: str
    device_priv_b64_enc: str
    device_priv_enc_method: str = "none"  # "none" | "passkey-prf" | "mnemonic-sealed"
    seed_b64: str | None = None  # BIP-39 seed bytes; None for ephemeral
    # Mnemonic words persisted ONLY when the user passes `tn init
    # --keep-mnemonic` — opt-in because storing the recovery phrase on
    # disk increases blast radius on a filesystem compromise. Off by
    # default. When present, `tn wallet export-mnemonic` can re-display.
    mnemonic_stored: str | None = None
    linked_vault: str | None = None
    prefs_version: int = 0
    prefs: IdentityPrefs = field(default_factory=IdentityPrefs)
    version: int = IDENTITY_SCHEMA_VERSION

    # Mnemonic words are shown ONCE at generation and never written.
    _mnemonic: str | None = field(default=None, repr=False, compare=False)
    _ephemeral: bool = field(default=False, repr=False, compare=False)
    _source_path: Path | None = field(default=None, repr=False, compare=False)

    @property
    def _seed(self) -> bytes | None:
        """In-memory seed for vault_wrap_key derivation.

        Sourced from seed_b64 (persisted) when loaded from disk. Old
        callers that set `ident._seed = ...` directly still work via
        the setter.
        """
        if self.seed_b64 is None:
            return None
        pad = "=" * (-len(self.seed_b64) % 4)
        return base64.urlsafe_b64decode(self.seed_b64 + pad)

    @_seed.setter
    def _seed(self, value: bytes | None) -> None:
        if value is None:
            self.seed_b64 = None
        else:
            self.seed_b64 = base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")

    # -- Factory: create fresh from a new mnemonic ------------------------

    @classmethod
    def create_new(cls, word_count: int = 12) -> Identity:
        """Generate a fresh mnemonic + derive everything from it.

        The returned Identity has `_mnemonic` populated; the caller is
        responsible for showing it to the user exactly once, then
        calling `ensure_written()` to persist everything except the
        mnemonic itself.
        """
        strength_bits = {12: 128, 15: 160, 18: 192, 21: 224, 24: 256}
        if word_count not in strength_bits:
            raise ValueError(f"word_count must be one of {list(strength_bits)}")
        m = Mnemonic("english")
        words = m.generate(strength=strength_bits[word_count])
        ident = cls.from_mnemonic(words)
        ident._mnemonic = words
        return ident

    # -- Factory: derive from a known mnemonic ---------------------------

    @classmethod
    def from_mnemonic(cls, words: str, *, passphrase: str = "") -> Identity:
        """Deterministically derive identity from a BIP-39 mnemonic.

        Same words + same passphrase → same DID. This is the recovery path.
        """
        m = Mnemonic("english")
        if not m.check(words):
            raise IdentityError("invalid BIP-39 mnemonic (bad checksum)")
        seed = m.to_seed(words, passphrase=passphrase)
        root = _hkdf(seed, HKDF_INFO_ROOT)
        device_priv_raw = _hkdf(root, HKDF_INFO_DEVICE)
        priv = Ed25519PrivateKey.from_private_bytes(device_priv_raw)
        pub = priv.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )
        ident = cls(
            did=_did_key_from_ed25519_pub(pub),
            device_pub_b64=base64.urlsafe_b64encode(pub).rstrip(b"=").decode("ascii"),
            device_priv_b64_enc=base64.urlsafe_b64encode(device_priv_raw)
            .rstrip(b"=")
            .decode("ascii"),
            device_priv_enc_method="none",
        )
        ident._seed = seed
        return ident

    # -- Factory: derive the vault wrap key --------------------------------

    def vault_wrap_key(self) -> bytes:
        """32 bytes of AES-256 key for client-side sealing."""
        if self._seed is None:
            raise IdentityError(
                "vault_wrap_key requires seed (rederive via from_mnemonic or from_passkey_prf)",
            )
        root = _hkdf(self._seed, HKDF_INFO_ROOT)
        return _hkdf(root, HKDF_INFO_VAULT_WRAP)

    # -- Factory: ephemeral (no persistence) -------------------------------

    @classmethod
    def create_ephemeral(cls) -> Identity:
        """Random device key, no mnemonic, no persistence.

        Used when tn.init(yaml) runs with no identity.json present and
        the caller just wants a working DID for the current process.
        The key never touches disk; lives only in this Identity instance.
        """
        priv = Ed25519PrivateKey.generate()
        priv_raw = priv.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )
        pub = priv.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )
        ident = cls(
            did=_did_key_from_ed25519_pub(pub),
            device_pub_b64=base64.urlsafe_b64encode(pub).rstrip(b"=").decode("ascii"),
            device_priv_b64_enc=base64.urlsafe_b64encode(priv_raw).rstrip(b"=").decode("ascii"),
            device_priv_enc_method="none",
        )
        ident._ephemeral = True
        return ident

    # -- Device private key accessor ---------------------------------------

    def device_private_key_bytes(self) -> bytes:
        """Raw Ed25519 private key bytes."""
        if self.device_priv_enc_method != "none":
            raise IdentityError(
                f"device_priv stored with encryption "
                f"{self.device_priv_enc_method!r} — unwrap before use",
            )
        pad = "=" * (-len(self.device_priv_b64_enc) % 4)
        return base64.urlsafe_b64decode(self.device_priv_b64_enc + pad)

    # -- Persistence -------------------------------------------------------

    @classmethod
    def load(cls, path: Path | None = None) -> Identity:
        """Read identity.json. Raises IdentityError if missing/corrupt."""
        p = path or _default_identity_path()
        if not p.is_file():
            raise IdentityError(f"identity not found at {p}")
        try:
            doc = json.loads(p.read_text(encoding="utf-8"))
        except json.JSONDecodeError as e:
            raise IdentityError(f"identity.json parse error at {p}: {e}") from e
        if doc.get("version") != IDENTITY_SCHEMA_VERSION:
            raise IdentityError(
                f"identity schema version {doc.get('version')} != "
                f"{IDENTITY_SCHEMA_VERSION}; run `tn wallet restore` or "
                f"upgrade your identity.json manually",
            )
        ident = cls(
            did=doc["did"],
            device_pub_b64=doc["device_pub_b64"],
            device_priv_b64_enc=doc["device_priv_b64_enc"],
            device_priv_enc_method=doc.get("device_priv_enc_method", "none"),
            seed_b64=doc.get("seed_b64"),
            mnemonic_stored=doc.get("mnemonic_stored"),
            linked_vault=doc.get("linked_vault"),
            prefs_version=int(doc.get("prefs_version", 0)),
            prefs=IdentityPrefs.from_dict(doc.get("prefs")),
            version=int(doc.get("version", IDENTITY_SCHEMA_VERSION)),
        )
        ident._source_path = p
        if ident.mnemonic_stored:
            ident._mnemonic = ident.mnemonic_stored
        return ident

    @classmethod
    def load_or_ephemeral(cls, path: Path | None = None) -> Identity:
        """Load identity.json if present; otherwise return an ephemeral one.

        Used by `tn.init(yaml_path)` for dev-ergonomic scenarios where
        no identity has been set up yet.
        """
        p = path or _default_identity_path()
        if p.is_file():
            try:
                return cls.load(p)
            except IdentityError:
                raise
        return cls.create_ephemeral()

    def ensure_written(self, path: Path | None = None) -> Path:
        """Persist identity.json. Mkdir parents, chmod 0600."""
        if self._ephemeral:
            raise IdentityError(
                "refusing to persist an ephemeral identity; generate via "
                "Identity.create_new() instead",
            )
        p = path or _default_identity_path()
        p.parent.mkdir(parents=True, exist_ok=True)
        doc = {
            "version": self.version,
            "did": self.did,
            "device_pub_b64": self.device_pub_b64,
            "device_priv_b64_enc": self.device_priv_b64_enc,
            "device_priv_enc_method": self.device_priv_enc_method,
            "seed_b64": self.seed_b64,
            "mnemonic_stored": self.mnemonic_stored,
            "linked_vault": self.linked_vault,
            "prefs_version": self.prefs_version,
            "prefs": asdict(self.prefs),
        }
        tmp = p.with_suffix(".json.tmp")
        tmp.write_text(
            json.dumps(doc, indent=2, sort_keys=True),
            encoding="utf-8",
        )
        # Atomic rename on POSIX; Windows needs explicit replace.
        if p.exists():
            p.unlink()
        tmp.rename(p)
        # 0600 on POSIX. On Windows the XDG home is user-scoped anyway.
        if sys.platform != "win32":
            os.chmod(p, 0o600)
        self._source_path = p
        return p


# ---------------------------------------------------------------------------
# DID endpoint resolution
# ---------------------------------------------------------------------------

# In-process cache: did_str -> base URL string.
_did_endpoint_cache: dict[str, str] = {}


def _resolve_did_endpoint(did_str: str) -> str:
    """Derive the HTTP base URL for a vault service from a DID string.

    Supported DID methods:

    ``did:key:z...``
        The key is self-describing; no document to fetch. Transport URL
        comes from the ``TN_VAULT_DEFAULT_BASE`` environment variable
        (default ``http://localhost:8790``). Useful for local dev and
        tests that point at a local tnproto-org instance.

    ``did:web:<host>`` or ``did:web:<host>:<path:segments>``
        Fetch ``https://<host>/.well-known/did.json`` once per process,
        look for a ``service`` entry with ``type == "TnVaultEndpoint"``
        and use its ``serviceEndpoint``. Falls back to
        ``https://<host>`` when no matching service is found.
        The result is pinned in memory for the lifetime of the process
        so further calls are free.

    Raises ``ValueError`` for unsupported DID methods.
    """
    if did_str in _did_endpoint_cache:
        return _did_endpoint_cache[did_str]

    if did_str.startswith("did:key:"):
        base = os.environ.get("TN_VAULT_DEFAULT_BASE", "http://localhost:8790").rstrip("/")
        _did_endpoint_cache[did_str] = base
        return base

    if did_str.startswith("did:web:"):
        # did:web:host  or  did:web:host:path:segments
        # per the did:web spec, colons after the host encode path separators.
        parts = did_str[len("did:web:") :].split(":")
        host = parts[0]
        parts[1:] if len(parts) > 1 else [".well-known", "did.json"]

        # Well-known URL is always at /.well-known/did.json for the host.
        well_known = f"https://{host}/.well-known/did.json"

        try:
            import urllib.request

            with urllib.request.urlopen(well_known, timeout=5) as resp:
                doc = json.loads(resp.read())
            for svc in doc.get("service", []):
                if svc.get("type") == "TnVaultEndpoint":
                    endpoint = svc["serviceEndpoint"].rstrip("/")
                    _did_endpoint_cache[did_str] = endpoint
                    _log.info(
                        "_resolve_did_endpoint: did:web %s -> %s (from DID doc)",
                        host,
                        endpoint,
                    )
                    return endpoint
        except Exception as exc:  # noqa: BLE001 — preserve broad swallow; see body of handler
            _log.warning(
                "_resolve_did_endpoint: could not fetch DID doc for %s (%s) "
                "-- falling back to https://%s",
                did_str,
                exc,
                host,
            )

        # Fallback: assume the host itself is the vault base.
        fallback = f"https://{host}"
        _did_endpoint_cache[did_str] = fallback
        return fallback

    raise ValueError(
        f"_resolve_did_endpoint: unsupported DID method in {did_str!r}. Supported: did:key, did:web"
    )

File: ts-sdk/src/identity.ts
// Port of tn_proto/python/tn/identity.py — the per-user GLOBAL device
// identity stored at `$XDG_DATA_HOME/tn/identity.json` (or the platform
// equivalent), shared across every ceremony minted on this machine.
//
// Why global: Python's `tn init` seeds each ceremony's device key from this
// one identity, so all of a user's projects share a single device DID. That
// shared DID is what makes warm-attach work — once `tn account connect`
// mints the DID onto a vault account (and stamps `linked_account_id` here),
// a brand-new `tn init` reuses the same DID and can authenticate + attach
// without a browser.
//
// File schema is kept compatible with the Python writer (same field names
// for the device key + links) so a shared identity.json round-trips between
// the two CLIs. Unknown fields are preserved on save.

import { existsSync, mkdirSync, readFileSync, renameSync, rmSync, writeFileSync } from "node:fs";
import { homedir, platform } from "node:os";
import { dirname, join } from "node:path";
import { Buffer } from "node:buffer";

import { DeviceKey } from "./core/signing.js";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha2";
import { generateMnemonic, validateMnemonic, mnemonicToSeedSync } from "@scure/bip39";
import { wordlist as englishWordlist } from "@scure/bip39/wordlists/english";

// HKDF parameters — byte-identical to python/tn/identity.py (_hkdf + HKDF_INFO_*):
// HKDF-SHA256, salt "tn:v1", info "tn:<stage>:v1", 32-byte output.
const HKDF_SALT = new TextEncoder().encode("tn:v1");
const HKDF_INFO_ROOT = new TextEncoder().encode("tn:root:v1");
const HKDF_INFO_DEVICE = new TextEncoder().encode("tn:device:v1");
const HKDF_INFO_VAULT_WRAP = new TextEncoder().encode("tn:vault:wrap:v1");

/** Resolve the identity directory. Mirrors Python `_default_identity_dir`. */
export function defaultIdentityDir(): string {
  const override = process.env["TN_IDENTITY_DIR"];
  if (override) return override;
  const xdg = process.env["XDG_DATA_HOME"];
  if (xdg) return join(xdg, "tn");
  if (platform() === "win32") {
    const base = process.env["APPDATA"] ?? join(homedir(), "AppData", "Roaming");
    return join(base, "tn");
  }
  return join(homedir(), ".local", "share", "tn");
}

/** Path to the global identity.json. Mirrors Python `_default_identity_path`. */
export function defaultIdentityPath(): string {
  return join(defaultIdentityDir(), "identity.json");
}

function _b64urlEncode(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64url");
}

function _b64urlDecode(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, "base64url"));
}

/**
 * The machine-global TN device identity. Holds the device Ed25519 seed (so
 * every ceremony minted from it shares one DID) plus the remembered vault +
 * account bindings used to drive warm-attach.
 */
export class Identity {
  readonly did: string;
  /** 32-byte Ed25519 device seed. */
  readonly seed: Uint8Array;
  linkedVault: string | null;
  linkedAccountId: string | null;
  /** Path this identity was loaded from / will be written to. */
  private _path: string;
  /** Raw doc as loaded, so unknown fields survive a save round-trip. */
  private _raw: Record<string, unknown>;
  /** In-memory 64-byte BIP-39 seed (mnemonic-derived identities only), kept
   *  for vaultWrapKey(); null for random/loaded identities. */
  private _bip39Seed: Uint8Array | null = null;
  /** In-memory mnemonic words. Populated by {@link createNew}/{@link fromMnemonic}
   *  (shown to the user once) and on load from a persisted `mnemonic_stored`.
   *  NEVER written to disk unless the caller opts in via `mnemonicStored`.
   *  Mirrors Python `Identity._mnemonic`. */
  private _mnemonic: string | null = null;

  private constructor(args: {
    did: string;
    seed: Uint8Array;
    linkedVault: string | null;
    linkedAccountId: string | null;
    path: string;
    raw: Record<string, unknown>;
  }) {
    this.did = args.did;
    this.seed = args.seed;
    this.linkedVault = args.linkedVault;
    this.linkedAccountId = args.linkedAccountId;
    this._path = args.path;
    this._raw = args.raw;
  }

  /** Path this identity was loaded from. Mirrors Python's identity_path. */
  get path(): string { return this._path; }

  /**
   * Ceremony-mode preference. Mirrors Python Identity.prefs.default_new_ceremony_mode.
   * Defaults to "local" if not present in the file.
   */
  get prefs(): { defaultNewCeremonyMode: string } {
    const p = this._raw["prefs"] as Record<string, unknown> | undefined;
    return {
      defaultNewCeremonyMode:
        typeof p?.["default_new_ceremony_mode"] === "string"
          ? (p["default_new_ceremony_mode"] as string)
          : "local",
    };
  }

  /**
   * Version counter for account-pulled prefs. Mirrors Python Identity.prefs_version.
   * Defaults to 0 if not present.
   */
  get prefsVersion(): number {
    const v = this._raw["prefs_version"];
    return typeof v === "number" ? v : 0;
  }

  /** Build a VaultIdentity-compatible DeviceKey from the device seed. */
  deviceKey(): DeviceKey {
    return DeviceKey.fromSeed(this.seed);
  }

  /** The in-memory mnemonic words, or null for an identity loaded without a
   *  persisted `mnemonic_stored`. Show ONCE at generation; do not log.
   *  Mirrors Python `Identity._mnemonic`. */
  get mnemonic(): string | null {
    return this._mnemonic;
  }

  /** The persisted recovery phrase (`mnemonic_stored` in identity.json), or
   *  null when not opted in. Mirrors Python `Identity.mnemonic_stored`. */
  get mnemonicStored(): string | null {
    const v = this._raw["mnemonic_stored"];
    return typeof v === "string" ? v : null;
  }

  /** Persist (or clear) the recovery phrase in identity.json on the next
   *  `save()`. Opt-in via `tn init --keep-mnemonic`; storing the phrase on
   *  disk widens the blast radius of a filesystem compromise. */
  set mnemonicStored(words: string | null) {
    if (words === null) delete this._raw["mnemonic_stored"];
    else this._raw["mnemonic_stored"] = words;
  }

  /**
   * Generate a fresh mnemonic and derive a brand-new identity from it.
   * Byte-parity with Python `Identity.create_new(word_count)`: the returned
   * identity carries the mnemonic in-memory ({@link mnemonic}) so the caller
   * can show it once, then `save()` (the phrase is NOT persisted unless the
   * caller also sets {@link mnemonicStored}). `wordCount` selects entropy:
   * 12→128, 15→160, 18→192, 21→224, 24→256 bits.
   */
  static createNew(wordCount = 12, opts: { path?: string } = {}): Identity {
    const strengthBits: Record<number, number> = { 12: 128, 15: 160, 18: 192, 21: 224, 24: 256 };
    const bits = strengthBits[wordCount];
    if (bits === undefined) {
      throw new Error(`word_count must be one of ${JSON.stringify(Object.keys(strengthBits).map(Number))}`);
    }
    const words = generateMnemonic(englishWordlist, bits);
    return Identity.fromMnemonic(words, opts);
  }

  /**
   * Deterministically derive the device identity from a BIP-39 mnemonic.
   * Byte-identical to Python `Identity.from_mnemonic`: BIP-39 seed -> HKDF
   * `tn:root:v1` -> HKDF `tn:device:v1` -> the 32-byte Ed25519 device key.
   * Same words + passphrase => same DID (the recovery path). Rejects a bad
   * checksum, mirroring Python's `Mnemonic.check`.
   */
  static fromMnemonic(
    words: string,
    opts: { passphrase?: string; path?: string } = {},
  ): Identity {
    if (!validateMnemonic(words, englishWordlist)) {
      throw new Error("invalid BIP-39 mnemonic (bad checksum)");
    }
    const bip39Seed = mnemonicToSeedSync(words, opts.passphrase ?? "");
    const root = hkdf(sha256, bip39Seed, HKDF_SALT, HKDF_INFO_ROOT, 32);
    const deviceSeed = hkdf(sha256, root, HKDF_SALT, HKDF_INFO_DEVICE, 32);
    const dk = DeviceKey.fromSeed(deviceSeed);
    const id = new Identity({
      did: dk.did,
      seed: deviceSeed,
      linkedVault: null,
      linkedAccountId: null,
      path: opts.path ?? defaultIdentityPath(),
      // Persist the 64-byte BIP-39 seed in seed_b64 (matches Python) so the
      // vault-wrap key survives a save/load round-trip.
      raw: { seed_b64: Buffer.from(bip39Seed).toString("base64url") },
    });
    id._bip39Seed = bip39Seed;
    // Keep the words in-memory so `tn init` can show the banner / honour
    // --keep-mnemonic. Mirrors Python (create_new sets `_mnemonic`; the
    // mnemonic-file path sets it explicitly). NOT persisted unless the caller
    // opts in via `mnemonicStored`.
    id._mnemonic = words;
    return id;
  }

  /**
   * Re-derive the 32-byte AES-256 vault-wrap key from the BIP-39 seed.
   * Byte-identical to Python `Identity.vault_wrap_key()`. Requires a
   * mnemonic-derived identity (or one loaded with a 64-byte `seed_b64`).
   */
  vaultWrapKey(): Uint8Array {
    let seed = this._bip39Seed;
    if (seed === null) {
      const sb = this._raw["seed_b64"];
      if (typeof sb === "string" && sb) seed = _b64urlDecode(sb);
    }
    if (seed === null) {
      throw new Error(
        "vaultWrapKey requires the BIP-39 seed (re-derive via Identity.fromMnemonic)",
      );
    }
    const root = hkdf(sha256, seed, HKDF_SALT, HKDF_INFO_ROOT, 32);
    return hkdf(sha256, root, HKDF_SALT, HKDF_INFO_VAULT_WRAP, 32);
  }

  /** Read identity.json. Throws if missing or corrupt. */
  static load(path?: string): Identity {
    const p = path ?? defaultIdentityPath();
    if (!existsSync(p)) {
      throw new Error(`identity.json not found at ${p}`);
    }
    let doc: Record<string, unknown>;
    try {
      doc = JSON.parse(readFileSync(p, "utf8")) as Record<string, unknown>;
    } catch (e) {
      throw new Error(`identity.json parse error at ${p}: ${(e as Error).message}`, { cause: e });
    }
    const encMethod = String(doc["device_priv_enc_method"] ?? "none");
    if (encMethod !== "none") {
      throw new Error(
        `identity.json device key is stored with encryption ${JSON.stringify(encMethod)}; ` +
          `tn-js cannot unwrap it (use the Python CLI on this machine)`,
      );
    }
    const privB64 = doc["device_priv_b64_enc"];
    if (typeof privB64 !== "string" || !privB64) {
      throw new Error(`identity.json at ${p} is missing device_priv_b64_enc`);
    }
    const seed = _b64urlDecode(privB64);
    if (seed.length !== 32) {
      throw new Error(`identity.json device seed must be 32 bytes; got ${seed.length}`);
    }
    const id = new Identity({
      did: String(doc["did"] ?? DeviceKey.fromSeed(seed).did),
      seed,
      linkedVault: (doc["linked_vault"] as string | null) ?? null,
      linkedAccountId: (doc["linked_account_id"] as string | null) ?? null,
      path: p,
      raw: doc,
    });
    // Restore the in-memory mnemonic from a persisted phrase (--keep-mnemonic),
    // so `tn wallet export-mnemonic` can re-display it. Mirrors Python load:
    // `ident._mnemonic = ident.mnemonic_stored`.
    if (typeof doc["mnemonic_stored"] === "string") id._mnemonic = doc["mnemonic_stored"];
    return id;
  }

  /**
   * Load the global identity if present, else mint a fresh device key and
   * persist it. Mirrors the Python `tn init` "reuse identity or create one"
   * behaviour so every ceremony on this machine shares one device DID.
   */
  static loadOrMint(path?: string): Identity {
    const p = path ?? defaultIdentityPath();
    if (existsSync(p)) return Identity.load(p);

    // Mint a fresh BIP-39-backed identity, matching Python's
    // Identity.create_new(word_count=12) (12 words = 128 bits of entropy):
    // bip39 seed -> HKDF root -> device key, persisting the 64-byte bip39 seed
    // (via fromMnemonic -> save) so vaultWrapKey() derives the SAME key
    // material as Python and the identity is BIP-39-structured. The previous
    // random 32-byte seed produced an identity whose seed_b64 was the device
    // seed, yielding a vault-wrap key incompatible with Python's.
    const id = Identity.createNew(12, { path: p });
    id.save();
    return id;
  }

  /** Persist identity.json (atomic-ish: write tmp, replace). Preserves
   *  unknown fields loaded from an existing file. Mirrors Python
   *  `Identity.ensure_written`. */
  save(path?: string): string {
    const p = path ?? this._path;
    mkdirSync(dirname(p), { recursive: true });
    const dk = DeviceKey.fromSeed(this.seed);
    const privB64 = _b64urlEncode(this.seed);
    const doc: Record<string, unknown> = {
      ...this._raw,
      version: this._raw["version"] ?? 1,
      did: this.did,
      device_pub_b64: _b64urlEncode(dk.publicKey),
      device_priv_b64_enc: privB64,
      device_priv_enc_method: "none",
      seed_b64: this._raw["seed_b64"] ?? privB64,
      linked_vault: this.linkedVault,
      linked_account_id: this.linkedAccountId,
    };
    // Sort TOP-LEVEL keys for deterministic output (mirrors Python's
    // sort_keys=True) WITHOUT an array replacer — an array replacer would
    // recursively filter out any nested key not in the list (e.g. prefs.*).
    const sorted: Record<string, unknown> = {};
    for (const k of Object.keys(doc).sort()) sorted[k] = doc[k];
    const tmp = `${p}.tmp`;
    writeFileSync(tmp, JSON.stringify(sorted, null, 2), "utf8");
    if (existsSync(p)) rmSync(p);
    renameSync(tmp, p);
    this._path = p;
    this._raw = doc;
    return p;
  }
}
